daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala

```diff
@@ -956,4 +1013,33 @@ private[validation] object Typing {
     case TypeConApp(tcon, targs) => targs.foldLeft[Type](TTyCon(tcon))(TApp)
   }
 
+  private[this] class ExpectedPatterns(val number: Int, patterns: => Iterator[CasePat]) {
+    def missingPatterns(ranks: Set[Int]): List[CasePat] =
+      patterns.zipWithIndex.collect { case (p, i) if !ranks(i) => p }.toList
+  }
+  private[this] object ExpectedPatterns {
+    def apply(patterns: CasePat*) = new ExpectedPatterns(patterns.length, patterns.iterator)
+  }
+
+  private[this] val wildcard: ExprVarName = Name.assertFromString("_")
+  private[this] def variantExpectedPatterns(
+      scrutTCon: TypeConName,
+      cons: ImmArray[(VariantConName, _)],
+  ) = new ExpectedPatterns(
+    cons.length,
+    cons.iterator.map { case (variants, _) => CPVariant(scrutTCon, variants, wildcard) },
```

Comment:
My understanding is that we only run through this iterator if we already know that we need to report an error. Is that correct?

Reply:
Yes. The argument `patterns` of `ExpectedPatterns` is "by-name" (See line 1016), so we will actually construct this iterator, only when we will call  the method `missingPatterns` of `ExpectedPatterns`. As you can see we do that only when we need to report an error.